armory/instrument/instrument.py

```diff
@@ -0,0 +1,552 @@
+        self.clear()
+        self._results = []
+        self.writers = []
+        self.auto_measure = bool(auto_measure)
+        self._warned = False
+        self.keep_results = bool(keep_results)
+        # TODO: final metric (mean) [and global metrics]
+        # TODO: do we need to handle metric kwargs?
+        # TODO:
+        # TODO: Add some sort of sampling rate (every batch, every x iterations, etc.)
+
+    def get_arg_names(self):
+        return list(self.arg_index)
+
+    def clear(self):
+        """
+        Clear all of the current values
+        """
+        self.values = [None] * len(self.arg_index)
+        self.values_set = [False] * len(self.arg_index)
+        self.batches = [None] * len(self.arg_index)
+
+    def add_writer(self, writer):
+        """
+        Add a writer to emit outputs to.
+        """
+        if writer not in self.writers:
+            self.writers.append(writer)
+
+    def set(self, name, value, batch):
+        """
+        Set the value for the arg name
+        """
+        if name not in self.arg_index:
+            raise ValueError(f"{name} not a valid arg name from {self.arg_index}")
+        i = self.arg_index[name]
+        self.values[i] = value
+        self.values_set[i] = True
+        self.batches[i] = batch
+        if self.auto_measure and self.is_ready():
+            self.measure()
+
+    def is_ready(self, raise_error=False):
+        """
+        Return True if all values have been set and batch numbers match
+            if raise_error is True, raise ValueError instead of returning False
+        """
+        if not all(self.values_set):
+            if raise_error:
+                raise ValueError(f"Not all values have been set: {self.values_set}")
+            return False
+        if any(self.batches[0] != batch for batch in self.batches):
+            if raise_error:
+                raise ValueError("Batch numbers are mismatched: {self.batches}")
+            return False
+        return True
+
+    def measure(self, clear_values=True):
+        self.is_ready(raise_error=True)
+        result = self.metric(*self.values)
+        record = (self.name, self.batches[0], result)
+        if self.keep_results:
+            self._results.append(record)
+        for writer in self.writers:
+            writer.write(record)
+        if not self.keep_results and not self.writers and not self._warned:
+            log.warning(
+                f"Meter {self.name} has no writer added and keep_results is False"
+            )
+            self._warned = True
+        if clear_values:
+            self.clear()
+
+    def finalize(self):
+        """
+        Primarily intended for metrics of metrics, like mean or stdev of results
+        """
+        pass  # TODO
+        # raise NotImplementedError()
+
+    def results(self):
+        if not self.keep_results:
+            raise ValueError("keep_results is False")
+        return copy.deepcopy(self._results)
```

Comment:
I find gratuitous copies cause more trouble than they avert. Since the test rig doesn't call this method, I can't tell the motivation for this.

Reply:
Yeah, probably good to remove.

Reply:
Removed.